Deconvolution3D

gvtulder opened the pull request:
This patch should provide a 3D equivalent of Deconvolution2D.
(See also #3773, #3981 and #4238.)

barrykui commented:
I tested it, failed with error below:

```
>>> model = Sequential()
>>> model.add(Deconvolution3D(3, 3, 3, 3, output_shape=(None, 3, 14, 14, 14), border_mode='valid', input_s
hape=(3, 12, 12, 12)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/keras/models.py", line 280, in add
    layer.create_input_layer(batch_input_shape, input_dtype)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/topology.py", line 370, in create_input_layer
    self(x)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/topology.py", line 514, in __call__
    self.add_inbound_node(inbound_layers, node_indices, tensor_indices)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/topology.py", line 572, in add_inbound_node
    Node.create_node(self, inbound_layers, node_indices, tensor_indices)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/topology.py", line 149, in create_node
    output_tensors = to_list(outbound_layer.call(input_tensors[0], mask=input_masks[0]))
  File "/usr/local/lib/python2.7/dist-packages/keras/layers/convolutional.py", line 1413, in call
    filter_shape=self.W_shape)
  File "/usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py", line 1814, in deconv3
d
    padding=padding)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/nn_ops.py", line 345, in conv3d_trans
pose
    output_shape_ = ops.convert_to_tensor(output_shape, name="output_shape")
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 657, in convert_t
o_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/constant_op.py", line 180, in _
constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/constant_op.py", line 163, in c
onstant
    tensor_util.make_tensor_proto(value, dtype=dtype, shape=shape))
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/tensor_util.py", line 422, in m
ake_tensor_proto
    tensor_proto.string_val.extend([compat.as_bytes(x) for x in proto_values])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/compat.py", line 64, in as_bytes
    (bytes_or_text,))
TypeError: Expected binary or unicode string, got (None, 14, 14, 14, 3)
```

gvtulder commented:
@barrykui: That's a TensorFlow problem. Instead of None as the first element of `output_shape` you have to specify the batch size. None works for Theano but TensorFlow doesn't like shapes with None, apparently. You'd have the same problem with `Deconvolution2D`.
It's a bit weird that you have to pass the batch size for `output_shape` but not for `input_shape`, but apparently that was a [deliberate choice for the Deconvolution2D interface](https://github.com/fchollet/keras/pull/3838#issuecomment-249426693).

gvtulder commented:
The first of these two new patches derive the batch size from the input, so that None should work with TensorFlow as well. The second patch will automatically add None to the `output_shape` if you didn't specify a batch size (it makes the `output_shape` behave similarly to the `input_shape` parameter).

gvtulder commented on keras/layers/convolutional.py:
```diff
@@ -1247,6 +1250,189 @@ def get_config(self):
+
+        # apply a 3x3x3 transposed convolution with stride 2x2x2 and 3 output filters on a 12x12x12 volume:
+        model = Sequential()
+        model.add(Deconvolution2D(3, 3, 3, 3, output_shape=(None, 3, 25, 25, 25), subsample=(2, 2, 2), border_mode='valid', input_shape=(3, 12, 12, 12)))
```
Indeed, good catch. Fixed now.

thefiddler commented:
@gvtulder: the batch_size inference bug should be fixed in tensorflow >= 0.12rc0

gvtulder commented:
@thefiddler, that's good to know, thanks. It's probably best to keep the explicit batch size for a while, I think, since it's not expensive.

fchollet commented on keras/backend/tensorflow_backend.py:
```diff
@@ -1554,9 +1554,11 @@ def in_top_k(predictions, targets, k):
 # CONVOLUTIONS
 
-def _preprocess_deconv_output_shape(shape, dim_ordering):
+def _preprocess_deconv_output_shape(x, shape, dim_ordering):
```
We should avoid this change of API.

fchollet commented on keras/backend/tensorflow_backend.py:
```diff
@@ -1554,9 +1554,11 @@ def in_top_k(predictions, targets, k):
-        shape = (shape[0], shape[2], shape[3], shape[1])
+        shape = (shape[0],) + tuple(shape[2:]) + (shape[1],)
+    if shape[0] is None:
+        shape = (tf.shape(x)[0],) + tuple(shape[1:])
```
Isn't this corner case supposed to be fixed at this point?

gvtulder replied:
Not in the current TensorFlow version (0.12.0-rc1).

fchollet commented on keras/layers/convolutional.py:
```diff
@@ -1247,6 +1250,189 @@ def get_config(self):
+        print(preds.shape)
+        # Theano GPU: (None, 3, 13, 13, 13)
+        # Theano CPU: (None, 3, 14, 14, 14)
+        # TensorFlow: (None, 14, 14, 14, 3)
```
These examples are confusing. dim ordering should be referred to as "th dim ordering / tf dim ordering". Please pick one of them and have the entire example use the same convention. You may then follow up with the same example in the alternative dim ordering.

gvtulder replied:
That should be fixed now (for deconv2d as well).

gvtulder commented:
Rebased to the current master to include the recent `filter_shape` fix. The issues you mentioned should be fixed now.

I've also changed the calls to Theano's `AbstractConv2d_gradInputs` to call to the `conv2d_grad_wrt_input` wrapper function. This seems cleaner and both interfaces were introduced in the same Theano release (0.8.0), so it shouldn't be a problem to replace them.

@thefiddler: I upgraded to TensorFlow 0.12.1-rc1, but that still seems to have problems with a None dimension for the batch size in the output_shape.

gvtulder commented:
Linking this to #4933, which proposes a similar fix for batch size `None`.

gvtulder commented:
@fchollet: Just a friendly reminder/question. This has been lying around for some time, so I was wondering if it makes sense to keep updating this. Is this still something you want to include? Or might want to include in the future but not now?

tboquet commented:
@gvtulder you could consider opening another PR (in parallel) in the [contrib repo](https://github.com/farizrahman4u/keras-contrib) :tada: ! There are more official reviewers so you could have more feedback.

gattia commented:
Hi All, 

I'm really hoping to try out and take advantage of a 3d deconvolution layer in a segmentation task and this pull request is perfect for that. I'm hoping for someones expertise/opinion. 

While waiting for this request to be accepted and available with the regular install of Keras I'm assuming I could pull the f-deconvolution3d branch of keras and build from source? I've historically installed keras with pip/conda but I'm assuming that following the instructions in the /docker ReadMe file would suffice. Or, would it be advised to merge these changes with another local version of keras that is more upto date and then build. 

Any advice would be greatly appreciated. 

Thanks, 

Anthony.

gvtulder commented:
@gattia: I just pushed an updated version of the patch, so you can use it with the current Keras master branch.

gattia commented:
@gvtulder Thanks for the clarification, and for this contribution. 

Anthony.

gattia commented:
I just wanted to make a note for anyone wanting to use Deconvolution 3D while it's still not on the official Keras. It has been accepted at the keras-contrib.  I personally didnt know what that meant. So to clarify for others, it means that you can install keras-contrib using pip and can use additional more "experimental" portions of keras (including Deconvolution 3D) before they are implemented in the full version. Installation instructions can be found @ https://github.com/farizrahman4u/keras-contrib

gattia commented:
Hi @gvtulder, 

Im hoping that Im missing something simple, but there seems to be an issue with the deconvolution layer. 

I tried it as the next layer in a network that I was working on but it failed. So, I tried using the examples in the convolution.py code and am still getting a similar issue. 

The function fails when you try to use a kernel that's 3rd dimension is not the same as the number of input channels, which shouldnt be a reason for it to fail. 

I've outlined a few steps that I tried to figure out when it works/doesnt work in a jupyter/ipython notebook. The notebook with my comments and some code is on my github account  [https://github.com/gattia/problemSolving/blob/master/deconv3d_testing.ipynb](url). 

Let me know if you need any information. 

Thanks, 

Anthony.

gvtulder commented:
Hi @gattia. Thanks, there is indeed something wrong with the order of the filter dimensions. The function in the TensorFlow backend is mixing the spatial dimensions with the channels. This should be fixed by the new patch, so if you apply that you should no longer see the first error in your example. (This was only a problem in the TensorFlow backend.)

The second error is a mistake in your example, I think.
```
Number of cols of out_backprop doesn't match computed: actual = 12, computed = 13
```
If you change the kernel shape, you'll also have to change the input shape or the output shape. If you apply a deconvolution with a 2-pixel filter to a 12-pixel input, the output will be 12+2-1=13, not 14.

gattia commented:
Hi @gvtulder, Thanks for that fix. I just pulled the new keras_contrib build and tried it all out. Everything is working as expected! 

This has been a great contribution, I really appreciate the work! Hopefully it makes its way into the regular Keras build soon. Ive update the [github repsitory/file](https://github.com/gattia/problemSolving/blob/master/deconv3d_testing.ipynb  ) that I used to test this, in case anyone is interested. 

Anthony.

fchollet commented:
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
